### ISSUE TITLE
NewRule: Local functions should use parameters instead of closure

### ISSUE THREAD

**Issue by viceroypenguin** [meziantou/Meziantou.Analyzer#333]
Similar to MA0105 where lambdas are detected as using a closure, it would be nice to have an analyzer for detecting local methods that are using a closure.

**Comment by meziantou** [comment]
Could you provide a sample to explain the rule you want? (valid and invalid code)

**Comment by viceroypenguin** [comment]
Sure! Key difference between the two below is that in the valid one, it does not bring `param` or `str1` into the closure. 

Valid:
```csharp
void Method(int param)
{
  string str1 = "test";
  Inner(param, str1);

  void Inner(int param, string str1)
  {
    Console.WriteLine(param);
    Console.WriteLine(str1);
  }
}
```

Invalid:
```csharp
void Method(int param)
{
  string str1 = "test";
  Inner();

  void Inner()
  {
    Console.WriteLine(param);
    Console.WriteLine(str1);
  }
}
```

**Comment by meziantou** [comment]
What you suggest is different from MA0105. MA0105 is about bad usages of `ConcurrentDictionary` (and it only detects this specific case for closures because you can fix it easily).
Here you are suggesting detecting potential allocations because of closures whatever the context.

- Visual Studio shows captured variables:

    ![image](https://user-images.githubusercontent.com/509220/173733347-a4ffcc16-8c22-43b9-a4e1-af438b4e8904.png)

- If you care about allocations for performance reasons, you can use the `static` keyword to prevent the introduction of a closure. 
- Also, I think the analyzer `Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers` ([rules](https://github.com/dotnet/roslyn-analyzers/blob/main/src/PerformanceSensitiveAnalyzers/Microsoft.CodeAnalysis.PerformanceSensitiveAnalyzers.md)) would be a better fit for this rule.

**Comment by viceroypenguin** [comment]
Ah, apologies - I didn't look into the details of MA0105. Since you're not doing this already, I withdraw my request; I thought it would be easy for you to adapt from one to the other.

Yes, there are IDE hints to help, but I've got about 70 methods I'm currently reviewing, so having the analyzer would make it easier to confirm I was able to get all of them.
